File: leetcode/js/two_sum_ii.js
/**
 * @param {number[]} numbers
 * @param {number} target
 * @return {number[]}
 */

// solution1
var twoSum = function(numbers, target) {
  let i = 0;
  let j = numbers.length - 1;
  const ret = [];
  while (i < j) {
    const sum = numbers[i] + numbers[j];
    if (sum < target) {
      i++;
    } else if (sum > target) {
      j--;
    } else {
      ret.push(i+1, j+1);
      break;
    }
  }
  return ret;
};

// solution2
var twoSum = function(numbers, target) {
  const map = new Map();
  const ret = [];
  numbers.every((num, index) => {
    if (map.has(target - num)) {
      ret.push(map.get(target - num) + 1, index + 1);
      return false;
    }
    map.set(num, index);
    return true;
  });
  return ret;
};

File: leetcode/java/two_sum_II.java
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        int[] ret = new int[2];
        int p = 0;
        int q = numbers.length - 1;
        while (p < q) {
            int sum = numbers[p] + numbers[q];
            if (sum == target) {
                ret[0] = p + 1;
                ret[1] = q + 1;
                break;
            }
            if (sum < target)
                p++;
            else
                q--;
        }
        return ret;
    }
}
